tests/unit/replication/ReplicationTaskScheduler.spec.js

```diff
@@ -1,39 +1,294 @@
-    '../../../extensions/replication/utils/ReplicationTaskScheduler');
-
-describe('replication task scheduler', () => {
-    it('should ensure serialization of updates to the same versioned object',
-    done => {
-        const taskScheduler = new ReplicationTaskScheduler(
-            (entry, done) => {
-                setTimeout(() => {
-                    // eslint-disable-next-line no-param-reassign
-                    entry.object.value = entry.setValueTo;
-                    done();
-                }, Math.random() * 100);
-            });
-        const objects = [];
-        let doneCount = 0;
+const ObjectQueueEntry = require('../../../lib/models/ObjectQueueEntry');
+const ActionQueueEntry = require('../../../lib/models/ActionQueueEntry');
+
+const ReplicationTaskScheduler =
+    require('../../../extensions/replication/utils/ReplicationTaskScheduler');
+
+function getMockQueueEntry(params) {
+    const { objectKey, versionId, contentMD5 } = params;
+    return new ObjectQueueEntry('test-bucket-name', objectKey)
+        .setVersionId(versionId)
+        .setContentMd5(contentMD5 || 'd41d8cd98f00b204e9800998ecf8427e');
+}
+
+function getMockActionEntry(params) {
+    const { objectKey, versionId, contentMD5 } = params;
+    return new ActionQueueEntry({
+        target: {
+            key: `test-bucket-name/${objectKey}`,
+            version: versionId,
+            contentMd5: contentMD5 || 'd41d8cd98f00b204e9800998ecf8427e',
+        }
+    });
+}
+
+function processingFunc(entry, cb) {
+    setTimeout(() => {
+        // eslint-disable-next-line no-param-reassign
+        entry.object.value = entry.setValueTo;
+        cb();
+    }, Math.random() * 100);
+}
+
+const replicationTaskScheduler = new ReplicationTaskScheduler(processingFunc);
+
+function assertRunningTasksAreCleared() {
+    assert.deepStrictEqual(
+        Object.keys(replicationTaskScheduler._runningTasksByMasterKey), []);
+    assert.deepStrictEqual(
+        Object.keys(replicationTaskScheduler._runningTasks), []);
+}
+
+const queueEntryClasses = [
+    'ObjectQueueEntry',
+    'ActionQueueEntry',
+];
+
+queueEntryClasses.forEach(mockClass => {
+    let results = {};
+    let objects = [];
+    let doneCount = 0;
+    let queueEntryClass;
+
+    function getMockEntry(params) {
+        if (queueEntryClass === 'ObjectQueueEntry') {
+            return getMockQueueEntry(params);
+        }
+        if (queueEntryClass === 'ActionQueueEntry') {
+            return getMockActionEntry(params);
+        }
+        return undefined;
+    }
+
+    function scheduleEntries(params, cb) {
+        const { uniqueKeyCount, hasUniqueVersions, hasUniqueContent } = params;
+        const duplicateKeyCount = params.duplicateKeyCount + 1;
+        for (let i = 0; i < uniqueKeyCount; ++i) {
+            const objectKey = `key_${i}`;
+            let versionId = uuid();
+            let contentMD5 = uuid();
+            for (let j = 0; j < duplicateKeyCount; ++j) {
```

Comment:
change to `j <= duplicateKeyCount`?

Reply:
I've added 1 to the value, since it's used below as well: https://github.com/scality/backbeat/pull/644/files/592eb2d9047720c614b25198e51b159ed0fe9e1c#diff-016c440d4159f39e05ea96d09aaaf274R68

Reply:
Ah, ok, I would recommend though to avoid mixing params and local variables with the same name but a different meaning, as it brings confusion, you sure can use a local variable that has the same value as the param but a shorter form for convenience, and set `duplicateKeyCount + 1` in the loop or use `<=` instead.

Reply:
Yeah, I agree that would improve readability. Updated!